### PR TITLE
Separate Map and Trajectory options from Node options

### PR DESCRIPTION
**Pull request by yutakaoka**
I separated Map and Trajectory options from Node options for now. After this PR, I will send another PR about multibot mapping node. Please review it.

### CODE REVIEW

**Comment by SirVer** [comment]
Sorry, I was quite loaded lately. Will review tomorrow!

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc [hunk 1]
```diff
@@ -64,18 +66,21 @@ void MapBuilderBridge::WriteAssets(const string& stem) {
     LOG(WARNING) << "No data was collected and no assets will be written.";
   } else {
     LOG(INFO) << "Writing assets with stem '" << stem << "'...";
-    if (options_.map_builder_options.use_trajectory_builder_2d()) {
+    CHECK_EQ(trajectory_options_.count(0), 1);
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Please for now add a TODO to support multi trajectory. I opened #343 to discuss what we want to do here.

File: cartographer_ros/cartographer_ros/trajectory_options.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+               (options.num_point_clouds > 0),
+           1)
+      << "Configuration error: 'use_laser_scan', "
+         "'use_multi_echo_laser_scan' and 'num_point_clouds' are "
+         "mutually exclusive, but one is required.";
+
+  return options;
+}
+
+TrajectoryOptions CreateTrajectoryOptionsWithTopics(
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/trajectory_options.cc]: This is unused and Lua files are not the correct place for ROS topics anyways. remapping is the way to go. Remove?

File: cartographer_ros/cartographer_ros/trajectory_options.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      trajectory_builder_options;
+  string tracking_frame;
+  string published_frame;
+  string odom_frame;
+  bool provide_odom_frame;
+  bool use_odometry;
+  bool use_laser_scan;
+  bool use_multi_echo_laser_scan;
+  int num_point_clouds;
+  string scan_topic;
```
Comment (SirVer) [inline cartographer_ros/cartographer_ros/trajectory_options.h]: Topics are usually hard coded and follow some convention which are established through topic remapping (like num_point_clouds). I am unsure how you envision the launch files and interfaces for multi robot to look, but putting topics into Lua files seems unconventional. Instead maybe the RPC that adds a new trajectory should contain the topic names?

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc [hunk 4]
```diff
@@ -23,24 +23,26 @@
...
   LOG(INFO) << "Added trajectory with ID '" << trajectory_id << "'.";
 
   CHECK_EQ(sensor_bridges_.count(trajectory_id), 0);
   sensor_bridges_[trajectory_id] =
       cartographer::common::make_unique<SensorBridge>(
-          tracking_frame, options_.lookup_transform_timeout_sec, tf_buffer_,
+          trajectory_options.tracking_frame,
+          map_options_.lookup_transform_timeout_sec, tf_buffer_,
           map_builder_.GetTrajectoryBuilder(trajectory_id));
+  trajectory_options_[trajectory_id] = trajectory_options;
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Done

File: cartographer_ros/cartographer_ros/map_builder_bridge.cc (hunk 1 above)
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/map_builder_bridge.cc]: Done

File: cartographer_ros/cartographer_ros/map_options.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+#include <string>
+
+#include "cartographer/common/lua_parameter_dictionary.h"
+#include "cartographer/common/port.h"
+#include "cartographer/mapping/map_builder.h"
+#include "cartographer_ros/sensor_bridge.h"
+
+namespace cartographer_ros {
+
+struct MapOptions {
```
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/map_options.h]: Now I am working on making "multibot_node_main" which needs this struct. I am thinking when the node boots up, it will make map_builder with this map options, and a ros service of starting trajectory will send a trajectory_options and start a new trajectory.

File: cartographer_ros/cartographer_ros/trajectory_options.cc (hunk 2 above)
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/trajectory_options.cc]: Removed

File: cartographer_ros/cartographer_ros/trajectory_options.h (hunk 3 above)
Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/trajectory_options.h]: OK, I will try to add the topic names into the ros service. Now I am thinking the service has contents of trajectory_options.

File: cartographer_ros/cartographer_ros/map_options.h (hunk 5 above)
Comment (SirVer) [inline cartographer_ros/cartographer_ros/map_options.h]: What is the semantic difference between multibot_node_main and node_main? Why are you not extending node_main to be multi-trajectory aware?

Comment (yutakaoka) [inline cartographer_ros/cartographer_ros/map_options.h]: I tried to extend node_main to be multi-trajectory aware, and send another PR which includes this commits. So this PR could be closed.

**Comment by SirVer** [comment]
Closed in favor of #346.
